### ISSUE TITLE
G005 "no implicit conditions" failures should report the implicit condition expression

### ISSUE THREAD

**Issue by bpfoley** [B-Lang-org/bsc#190]
A trivial example
```
-- bsc -sim  Test.bs
package Test() where

import FIFO

{-# properties mkModule1 = { verilog } #-} 
mkModule1 :: (IsModule m mType) => m Empty
mkModule1 = module
  f :: FIFO Bool <- mkFIFO
  rules
    {-# ASSERT no implicit conditions #-}
    "rule1": when False ==> do
      f.deq
```

Running `bsc -sim  Test.bs` reports

```
Error: "Test.bs", line 12, column 4: (G0005)
  The assertion `no_implicit_conditions' failed for rule `rule1'
  During elaboration of rule `rule1' at "Test.bs", line 12, column 4.
  During elaboration of `mkModule1' at "Test.bs", line 7, column 0.
```

but particularly for longer rules with multiple actions (or actions composed using joinActions or similar), it would be helpful to report which implicit condition triggered the warning.

**Comment by quark17** [comment]
I agree that it would help to print this information.  Some things to consider:

- Printing of ISyntax may not be very friendly, since it contains type arguments and other desugarings.  This shouldn't stop us, it just might mean more work (to improve the printing routines) is needed before adding this to the error message by default.  (Note that some error messages mention a flag that can be used to turn on more information in the error message; we could do something similar here, until we're ready to make it the default.)

- For BSV users, in particular, the current printing of expressions might be unhelpful, so make sure that you test this on BSV programs as well.

- The expression for the condition could get very large and not helpful to print in its entirety: If things go terribly awry, a complicated rule with for-loops etc could generate a huge expression; consider checking for overly large expressions and only printing the first few bits and eliding the rest -- again, possibly with a flag to turn on the full printing.

FYI, `G0005` is `ERuleAssertion`, which is used to report a number of different rule assertions.  The specific one for `no_implicit_conditions` is this code in `IExpand::evalRule`:
```
        -- traceM (ppReadable pp)                                                                                                   
	when ((RPnoImplicitConditions `elem` asrts) && not (isTrue pp)) $
            eAssertion rId RPnoImplicitConditions ""
```
where you can see that someone has added a trace of the expression, for debugging.  The second argument to the constructor is additional text that can be given in the message, so that could probably be used for reporting the condition -- although it might not give enough control over the formatting, for large expressions, so we might need to change that from a `String` to a `Doc`?
